pkg/transport/transport.go

```diff
@@ -15,43 +15,80 @@
 package transport
 
 import (
+	"context"
 	"net"
 	"net/http"
 	"strings"
 	"time"
 )
 
+type CancelableTransport struct {
+	*http.Transport
+	ctx    context.Context
+	cancel context.CancelFunc
+}
+
+func (c *CancelableTransport) Ctx() context.Context { return c.ctx }
+func (c *CancelableTransport) Cancel()              { c.cancel() }
+
+func (c *CancelableTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	if req.Context() != context.Background() {
+		return c.Transport.RoundTrip(req)
+	}
+	return c.Transport.RoundTrip(req.WithContext(c.ctx))
+}
+
 type unixTransport struct{ *http.Transport }
 
-func NewTransport(info TLSInfo, dialtimeoutd time.Duration) (*http.Transport, error) {
+func NewTransport(info TLSInfo, dialtimeoutd time.Duration) (*CancelableTransport, error) {
 	cfg, err := info.ClientConfig()
 	if err != nil {
 		return nil, err
 	}
 
+	ctx, cancel := context.WithCancel(context.TODO())
+
 	t := &http.Transport{
 		Proxy: http.ProxyFromEnvironment,
-		Dial: (&net.Dialer{
-			Timeout: dialtimeoutd,
-			// value taken from http.DefaultTransport
-			KeepAlive: 30 * time.Second,
-		}).Dial,
 		// value taken from http.DefaultTransport
 		TLSHandshakeTimeout: 10 * time.Second,
 		TLSClientConfig:     cfg,
 	}
+	ct := &CancelableTransport{
+		Transport: t,
+		ctx:       ctx,
+		cancel:    cancel,
+	}
+	tdialer := &net.Dialer{
+		Timeout: dialtimeoutd,
+		// value taken from http.DefaultTransport
+		KeepAlive: 30 * time.Second,
+	}
+	tdial := func(net, addr string) (net.Conn, error) {
+		if ctx.Err() != nil {
+			return nil, ctx.Err()
+		}
+		return tdialer.DialContext(ctx, net, addr)
```

Comment:
do you know if `DialContext()` handle cancelled/erred `ctx` properly? so that you don't have to check the `ctx.Err() != nil` beforehand.